src/integration/java/io/pinecone/integration/dataPlane/UpsertAndQueryServerlessTest.java

```diff
@@ -0,0 +1,238 @@
+
+            // Verify the correct vector id was updated
+            assert scoredVectorV1 != null;
+            assertEquals(scoredVectorV1.getId(), upsertIds.get(0));
+
+            // Verify the updated values
+            assertEquals(values, scoredVectorV1.getValuesList());
+
+            // Verify the updated metadata
+            assertEquals(scoredVectorV1.getMetadata(), metadataStruct);
+
+            // Verify the initial sparse values set for upsert operation
+            List<Long> unsignedIndicesList = new ArrayList<>(scoredVectorV1.getSparseValuesWithUnsignedIndices().getIndicesWithUnsigned32IntList());
+            Collections.sort(unsignedIndicesList);
+            Collections.sort(sparseIndices);
+            assertEquals(unsignedIndicesList, sparseIndices);
+
+            // Verify the initial sparse values set for upsert operation
+            List<Float> expectedSparseValues = new ArrayList<>(scoredVectorV1.getSparseValuesWithUnsignedIndices().getValuesList());
+            Collections.sort(expectedSparseValues);
+            Collections.sort(sparseValues);
+            assertEquals(expectedSparseValues, sparseValues);
+        });
+    }
+
+    @Test
+    public void upsertNullSparseIndicesNotNullSparseValuesSyncTest() {
+        String id = RandomStringBuilder.build(3);
+        StringBuilder exceptionMessage = new StringBuilder();
+        try {
+            index.upsert(id,
+                    generateVectorValuesByDimension(dimension),
+                    null,
+                    generateVectorValuesByDimension(dimension),
+                    null,
+                    null);
+        } catch (PineconeValidationException validationException) {
+            exceptionMessage.append(validationException.getLocalizedMessage());
+        } finally {
+            assertEquals(exceptionMessage.toString(), "Invalid upsert request. Please ensure that both sparse indices and values are present.");
+        }
+    }
+
+    @Test
+    public void upsertOptionalVectorsAndQueryIndexFutureTest() throws InterruptedException, ExecutionException {
+        int numOfVectors = 5;
+        DescribeIndexStatsResponse describeIndexStatsResponse1 = asyncIndex.describeIndexStats(emptyFilterStruct).get();
+        // Confirm the starting state by verifying the dimension of the index
+        assertEquals(describeIndexStatsResponse1.getDimension(), dimension);
+
+        // upsert vectors with required + optional parameters
+        List<String> upsertIds = getIdsList(numOfVectors);
+        int topK = 5;
+        String namespace = RandomStringBuilder.build("ns", 8);
+        List<Float> values = generateVectorValuesByDimension(dimension);
+        List<Long> sparseIndices = generateSparseIndicesByDimension(dimension);
+        List<Float> sparseValues = generateVectorValuesByDimension(dimension);
+        Struct metadataStruct = generateMetadataStruct();
+        for (String id : upsertIds) {
+            asyncIndex.upsert(id,
```

Comment:
I'm just realizing, our `upsert` operations in `Index` / `AsyncIndex` only support a single vector / id at a time, right?

The other clients allow passing an array of records:
- https://sdk.pinecone.io/typescript/classes/Index.html#upsert
- https://sdk.pinecone.io/python/pinecone.html#upsert-vectors

which we should definitely support here. I think this is a broader conversation though, I'm not sure what that looks like with the current generated classes.

Reply:
That's correct, I'll start a conversation on this internally.

Reply:
Added the upsert(List<Vector>, String namespace)